lib/src/main/java/io/ably/lib/types/MessageExtras.java

```diff
@@ -1,64 +1,150 @@
 package io.ably.lib.types;
 
 import com.google.gson.JsonElement;
-import io.ably.lib.util.Log;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
 import io.ably.lib.util.Serialisation;
-import org.msgpack.core.MessageFormat;
 import org.msgpack.core.MessagePacker;
 import org.msgpack.core.MessageUnpacker;
+import org.msgpack.value.ImmutableMapValue;
+import org.msgpack.value.ImmutableValue;
+import org.msgpack.value.Value;
+import org.msgpack.value.ValueFactory;
 
 import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Map;
+import java.util.Objects;
 
 public final class MessageExtras {
 	private static final String TAG = MessageExtras.class.getName();
-	
-	// TODO what about the 1.1 push extension? (TM2i)
 
-	private final DeltaExtras delta; // may be null
-	
+	private static final String DELTA = "delta";
+
+	private final DeltaExtras delta;
+	private final JsonObject raw;
+
+	/**
+	 * Creates a MessageExtras instance to be sent as extra with a Message to Ably's servers.
+	 *
+	 * @see <a href="https://www.ably.io/documentation/general/push/publish#channel-broadcast-example">Channel-based push notification example</a>
+	 *
+	 * @since 1.2.1
+	 */
+	public MessageExtras(final JsonObject raw) {
+		this.raw = raw;
+		delta = null;
+	}
+
+	/**
+	 * @since 1.2.0
+	 */
 	public MessageExtras(final DeltaExtras delta) {
```

Comment:
I'm still not clear on how this API is going to grow to accommodate more known `extras` members. We're going to have `push`, `metadata`, etc. How would you construct a `MessageExtras` if you have multiple of those values to supply?

The kind of thing you're seeing in other languages now is where the type for each of the known members implements a specific interface, eg

```
interface MessageExtra {}

class DeltaExtras implements MessageExtra {
  ...
}

class PushExtras implements MessageExtra {
  ...
}

public MessageExtras(JSONObject raw, MessageExtra ...members) { .... }
```

Reply:
Interesting. Depending on pressure we're under to release 1.2.1 it would definitely be worth having a call to discuss this [as you suggested elsewhere](https://ably-real-time.slack.com/archives/DPY0JSV6E/p1591911561044500).

The main thing I don't want to do is make any breaking changes between 1.2.0 and 1.2.1, but what you propose merely widens the type acceptance of this method so I think that's ok.

Reply:
@paddybyers I'm currently disinclined to do anything about this in order to get this fix out... we can think how to architect this properly outside of the scope of this PR, if that's ok with you?

In terms of what I [mentioned above](https://github.com/ably/ably-java/pull/581#discussion_r439088702)... If we added a super interface at some point in the future to `DeltaExtras` then this method could be safely refactored to accept any object that implements that interface without that being a breaking API change.

Reply:
> `public MessageExtras(final DeltaExtras delta)`

who will use this API?

Reply:
For some reason or another it looks like it's only [used by RealtimeDeltaDecoderTest](https://github.com/ably/ably-java/blob/b53dc120ef67999e6a48700fedf3bc5176eaa20f/lib/src/test/java/io/ably/lib/test/realtime/RealtimeDeltaDecoderTest.java#L148) at the moment. I'll have a look as to whether that's needed. Perhaps I can replace it with a mock.

Meanwhile, of course, do we simply ignore the fact that we exposed this for 1.2.0? Maybe the best option is to simply get rid of it from the public API. What do you think, @paddybyers?

(equally I guess you could follow that argument through and just revert this portion of [the change](https://github.com/ably/ably-java/commit/2b5c3a9ed0102d9b12803516ce8bf90ad15fa4d4) to the type of message `extras` wholesale!)

Reply:
I've removed that API: https://github.com/ably/ably-java/pull/581/commits/d6ff228bfb8c4c939ed63d5cf1f699092c8ed143